### ISSUE TITLE
Unable to update list of relationships

### ISSUE THREAD

**Issue by robin850** [spring-projects/spring-data-neo4j#2190]
Hello,

I have a basic relationship between server nodes where one node can have an open port for another server. Basically, something like this:

~~~java
@Node
public class Server {
    @Id @GeneratedValue
    private Long id;

    @Relationship("PORT_OPEN")
    private List<PortOpen> openPorts;
}

@RelationshipProperties
class PortOpen {
    @Id @GeneratedValue
    private Long id;

    @TargetNode
    private Server server;

    private Integer port;
}
~~~

I'm trying to update some existing relationships' target node:

~~~java
@Transactional
public void update(Server original, Server target) {
    List<Server> servers = serverRepository.havingOpenPortFor(original);
    
    // Update `servers`'s `openPorts` removing ones with `original` and creating
    // ones with `target`.

    serverRepository.saveAll(servers);
}
~~~

But it doesn't work ; previous relationships remain and new ones don't exist. Am I doing something wrong ? I tried with version 6.0.5 and 6.0.6.

Thank you very much for your help.

**Comment by michael-simons** [comment]
Hi. Thanks for using SDN.

This pattern works well for me:

```
	@Test
	void gh(@Autowired  Neo4jTemplate t, @Autowired TransactionTemplate transactionTemplate) {

		long id = transactionTemplate.execute(tx -> {
			Server s = new Server("s1");
			s.setOpenPorts(Arrays.asList(new PortOpen(new Server("s2"), 23)));
			return t.save(s).getId();
		});

		transactionTemplate.executeWithoutResult(tx -> {

			Server s = t.findById(id, Server.class).get();
			assertThat(s.getOpenPorts().removeIf(o -> o.getPort() == 23)).isTrue();
			s.getOpenPorts().add(new PortOpen(new Server("s3"), 4711));
			t.save(s);
		});
	}
```

Would you please share what's hidden behind the comment? Thank you.

**Comment by robin850** [comment]
Thanks for the quick reply. Sure, here's what's hidden behind the comment:

~~~java
servers.forEach(server -> {
    server.setOpenPorts(server.getOpenPorts()
            .stream()
            .map(entry -> {
                if (entry.getServer().getId().equals(original.getId()))
                    return new PortOpen(target, entry.getPort());
                else
                    return entry;
            }).collect(Collectors.toList()));
});
~~~

**Comment by robin850** [comment]
Trying to upgrade to SDN 6.0.7, there seems to be a difference with 6.0.7 : it works as long as there is a relationship between nodes just one way around.

Basically, the `update` method defined in the original message is "merging" two nodes into a single one. For example, having something like `(a)-[:PORT_OPEN]->(b)` and an untied node `c`, merging `b` and `c` works.

But, having something like `(a)-[:PORT_OPEN]->(b)` and `(c)-[:PORT_OPEN]->(a)`, merging `b` and `c` deletes the first relationship but doesn't save the `(a)->(c)` one.

**Comment by meistermeier** [comment]
@robin850 could you test `6.0.9-SNAPSHOT`? We have improved the handling in 6.0.8 and 6.0.9 in those cases a lot and it would really help to get your feedback.

**Comment by meistermeier** [comment]
Closing this because there were no response in the last three weeks.
Feel free to open it again if there is any addition.

**Comment by robin850** [comment]
Hello there !

I'm really sorry for not giving an answer sooner, I actually stopped working on the concerned project during a few months.

I couldn't manage to produce a failing test case and can't understand why. Actually, what I didn't mention originally is that right after calling the `update` method as described in the original post, there are some changes on the `target` variable that happen and a `save` call. Basically, the controller looks like this:

~~~java
myService.update(original, target);

myMapper.update(target);
serverRepository.save(target);
~~~

And actually, commenting the `save` call make the problem disappear.

Here is a Spring boot Application that reproduces the error : [demo.zip](https://github.com/spring-projects/spring-data-neo4j/files/7211389/demo.zip). The unit test is left even though I can't manage to get it failing.

**Warning** : By default, the application connects to a real Neo4J connection, not an embedded one.

To reset the nodes, you can run:

~~~
$ curl -X POST http://localhost:8080/servers
~~~

This will give you a JSON with the servers' ids. To run the `update` method, you can run the following, giving the JSON output by the `POST` request as the body:

~~~
$ curl -X PUT -H "Content-Type: application/json" -d '...' http://localhost:8080/servers
~~~

To switch between a working and a failing scenario, you just need to comment the `save` call in the `PUT` handler. I hope that's easy enough for you to debug. Let me know if you need/prefer something else.

Thank you very much for your time !

**Comment by meistermeier** [comment]
The underlying problem is that you are loading three different object trees. (Reduced for readability)
```
Server toUpdate = serverRepository.findById(dto.getToUpdate());
Server previous = serverRepository.findById(dto.getPrevious());
Server target = serverRepository.findById(dto.getTarget());
```

With the trees: 
```
1. toUpdate -port_open-> previous
2. previous (nothing)
3. target -port_open-> toUpdate
```
The business logic sets the ports in `toUpdate` (tree 1) to point towards `target` (tree 3).
But in tree 3 you have already loaded another `toUpdate` that still has the old ports definition.
After everything is persisted as wanted you call `save(target)` on tree 3 which will trigger also `toUpdate` to get updated (with the old information).

As you can see it is not really a bug in Spring Data Neo4j but a intentional behavior on the loading side that brings you in this situation. Both `toUpdate` objects are not the "same" in memory but the latter still has the outdated information because it were never modified.
There is no problem calling `save(toUpdate)` with the loaded `target` and its `toUpdate` because the single save operation checks if an object with a certain identifier was already processed and it will (coincidentally but reliable) skip the "outdated" information.

tl;dr; Spring Data Neo4j cannot determine (for your example) which information (`toUpdate`) object tells the truth and it will just execute the save operations after each other with the given object (and its related objects).

**Comment by robin850** [comment]
Oh gotcha, that's much clearer now, thanks for investigating and for the explanations ! Sorry I didn't understand it sooner.

Thank you very much for your time ! ❤️

**Comment by meistermeier** [comment]
Thank you for keep asking the question because it will help others.
Also it might be not intuitive on the first view but in the end it makes sense in the broader concept of Spring Data Neo4j, at least this is what we think.
